kikuchipy/indexation/similarity_metrics.py

```diff
@@ -0,0 +1,394 @@
+    |              |        returns        |       returns        |
+    +--------------+-----------+-----------+----------+-----------+
+    | MANY_TO_MANY | (n,m,y,x)    (N,y,x)  | (nm,yx)      (N,yx)  |
+    |              |        (N,n,m)        |        (N,nm)        |
+    +--------------+-----------+-----------+----------+-----------+
+    | ONE_TO_MANY  |   (y,x)      (N,y,x)  |   (yx,)      (N,yx)  |
+    |              |         (N,)          |        (N,)          |
+    +--------------+-----------+-----------+----------+-----------+
+    | MANY_TO_ONE  | (n,m,y,x)     (y,x)   | (nm,yx)      (yx,)   |
+    |              |         (n,m)         |        (nm,)         |
+    +--------------+-----------+-----------+----------+-----------+
+    | ONE_TO_ONE   |   (y,x)       (y,x)   |   (yx,)      (yx,)   |
+    |              |         scalar        |        scalar        |
+    +--------------+-----------+-----------+----------+-----------+
+    where (m,n) and (x,y) correspond to navigation and signal shape in HyperSpy, respectively.
+
+    Parameters
+    ----------
+    metric_func : Callable
+        Metric function with signature
+        `metric_func(patterns,templates)`
+
+        Computes similarity or distance matrix
+        between (experimental) pattern(s) and (simulated) template(s).
+
+    greater_is_better : bool, optional
+        Whether greater values correspond to more similar images, by default True.
+        Used for choosing `n_largest` metric results in `template_match`.
+    scope : MetricScope, optional
+        Describes how `metric_func`'s input parameters is structured,
+        by default MetricScope.MANY_TO_MANY.
+    flat : bool, optional
+        Whether patterns and templates are flattened before sent to `metric_func`
+        when the similarity metric is called, by default False.
+    make_compatible_to_lower_scopes : bool, optional
+        Whether to pad patterns and templates with single dimensions
+        to match given scope, by default False.
+    dtype_out : data-type, optional
+        The data type used and returned by the metric, by default np.float32
+
+    Returns
+    -------
+    Union[SimilarityMetric, FlatSimilarityMetric]
+        A callable class instance computing a similarity matrix with signature
+        `metric(patterns,templates)`.
+    """
+    sign = 1 if greater_is_better else -1
+    if flat:
+        return FlatSimilarityMetric(
+            metric_func,
+            sign,
+            scope,
+            flat,
+            make_compatible_to_lower_scopes,
+            dtype_out,
+        )
+    else:
+        return SimilarityMetric(
+            metric_func,
+            sign,
+            scope,
+            flat,
+            make_compatible_to_lower_scopes,
+            dtype_out,
+        )
+
+
+class SimilarityMetric:
+    """Similarity metric between 2D gray-tone images."""
+
+    _P_T_NDIM_TO_SCOPE = {
```

Comment:
You should here reference (in a comment or in the class docstring) the very nice table in `make_similarity_metric()`, so that the shapes are clear. The same goes for the `FlatSimilarityMetric` class.